src/allmydata/scripts/tahoe_backup.py

```diff
@@ -313,14 +263,353 @@ def upload(self, childpath):
             if bdb_results:
                 bdb_results.did_upload(filecap)
 
-            self.files_uploaded += 1
-            return filecap, metadata
+            return True, filecap, metadata
 
         else:
             self.verboseprint("skipping %s.." % quote_local_unicode_path(childpath))
-            self.files_reused += 1
-            return bdb_results.was_uploaded(), metadata
+            return False, bdb_results.was_uploaded(), metadata
 
 def backup(options):
     bu = BackerUpper(options)
     return bu.run()
+
+
+def collect_backup_targets(root, listdir, filter_children):
+    """
+    Yield BackupTargets in a suitable order for processing (deepest targets
+    before their parents).
+    """
+    try:
+        children = listdir(root)
+    except EnvironmentError:
+        yield PermissionDeniedTarget(root, isdir=True)
+    except FilenameEncodingError:
+        yield FilenameUndecodableTarget(root, isdir=True)
+    else:
+        for child in filter_children(children):
+            assert isinstance(child, unicode), child
+            childpath = os.path.join(root, child)
+            if os.path.islink(childpath):
+                yield LinkTarget(childpath, isdir=False)
+            elif os.path.isdir(childpath):
+                child_targets = collect_backup_targets(
+                    childpath,
+                    listdir,
+                    filter_children,
+                )
+                for child_target in child_targets:
+                    yield child_target
+            elif os.path.isfile(childpath):
+                yield FileTarget(childpath)
+            else:
+                yield SpecialTarget(childpath)
+        yield DirectoryTarget(root)
+
+
+
+def run_backup(
+        warn,
+        upload_file,
+        upload_directory,
+        targets,
+        start_timestamp,
+        stdout,
+):
+    progress = BackupProgress(warn, start_timestamp, len(targets))
+    for target in targets:
+        # Pass in the progress and get back a progress.  It would be great if
```

Comment:
I greatly prefer this style already (i.e. a generator). Easier to read the code! :+1: